server/task/src/main/scala/org/apache/james/task/eventsourcing/TerminationSubscriber.scala

```diff
@@ -0,0 +1,50 @@
+/** **************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ * *
+ * http://www.apache.org/licenses/LICENSE-2.0                   *
+ * *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ * ***************************************************************/
+
+package org.apache.james.task.eventsourcing
+
+import java.util.concurrent.{LinkedBlockingQueue, TimeUnit}
+
+import org.apache.james.eventsourcing.{Event, Subscriber}
+import org.reactivestreams.Publisher
+
+import reactor.core.publisher.Flux
+
+trait TerminationSubscriber extends Subscriber {
+  override def handle(event: Event): Unit = event match {
+    case event: Completed => addEvent(event)
+    case event: Failed => addEvent(event)
+    case event: Cancelled => addEvent(event)
+    case _ =>
+  }
+
+  def addEvent(event: Event): Unit
+
+  def listenEvents: Publisher[Event]
+}
+
+class MemoryTerminationSubscriber extends TerminationSubscriber {
+  private val events = new LinkedBlockingQueue[Event]()
+
+  override def addEvent(event: Event) =
+    events.add(event)
+
+  override def listenEvents: Publisher[Event] =
+    Flux.generate(sink => sink.next(events.poll(Long.MaxValue, TimeUnit.DAYS)))
```

Comment:
I used write Flux that way but I discovered Processors recently. If I understand correctly, they are made exactly for this purpose: having java code publish things into a Flux.
In you case, you probably want to read https://projectreactor.io/docs/core/release/reference/#_safely_produce_from_multiple_threads_by_using_the_sink_facade

> FluxProcessor sinks safely gate multi-threaded producers and can be used by applications that generate data from multiple threads concurrently. For example, a thread-safe serialized sink can be created for UnicastProcessor:
>
> ```
> UnicastProcessor<Integer> processor = UnicastProcessor.create();
> FluxSink<Integer> sink = processor.sink(overflowStrategy);
> ```
> Multiple producer threads may concurrently generate data on the following serialized sink:
>
> ```
> sink.next(n);
> ```

WDYT?